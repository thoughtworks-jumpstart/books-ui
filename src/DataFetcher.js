import React from "react";

const API_HOST = process.env.REACT_APP_BOOKS_API || "http://localhost:3000";

class DataFetcher extends React.Component {
  state = {
    data: [],
    isLoading: true
  };

  async componentDidMount() {
    const url = `${API_HOST}${this.props.endpoint}`;
    try {
      const response = await fetch(url);
      if (response.ok) {
        const data = await response.json();
        this.setState({
          data
        });
      }
    } catch (err) {
      console.err(err);
    } finally {
      this.setState({
        isLoading: false
      });
    }
  }

  render() {
    if (this.state.isLoading) {
      return <div>Loadingâ€¦</div>;
    }
    return this.props.render(this.state);
  }
}

export default DataFetcher;
