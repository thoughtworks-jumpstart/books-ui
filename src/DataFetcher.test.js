import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";
import DataFetcher, { API_HOST } from "./DataFetcher.js";

describe("when data is not loaded", () => {
  it("shows a loading message", () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <DataFetcher endpoint="/some-endpoint" render={jest.fn()} />
    );
    const output = renderer.getRenderOutput();
    expect(output).toMatchSnapshot();
  });
});

describe("when data is fetched", () => {
  let renderer;

  beforeEach(() => {
    fetch.resetMocks();
    renderer = new ShallowRenderer();
  });

  it("calls fetch with the api host and the supplied endpoint", async () => {
    fetch.mockResponseOnce(JSON.stringify([]));
    const endpoint = "/some-endpoint";
    renderer.render(<DataFetcher endpoint={endpoint} render={jest.fn()} />);
    const instance = renderer.getMountedInstance();
    await instance.componentDidMount();

    expect(fetch).toHaveBeenCalledWith(`${API_HOST}${endpoint}`);
  });

  it("calls the render prop", async () => {
    fetch.mockResponseOnce(JSON.stringify([]));
    const mockRenderProp = jest.fn();
    renderer.render(
      <DataFetcher endpoint="/some-endpoint" render={mockRenderProp} />
    );
    const instance = renderer.getMountedInstance();
    await instance.componentDidMount();

    expect(mockRenderProp).toHaveBeenCalled();
  });

  it("sets the fetched data in the state", async () => {
    const data = "I fetched some data!";
    fetch.mockResponseOnce(JSON.stringify(data));
    const mockRenderProp = jest.fn();
    renderer.render(
      <DataFetcher endpoint="/some-endpoint" render={mockRenderProp} />
    );
    const instance = renderer.getMountedInstance();
    await instance.componentDidMount();

    expect(instance.state.data).toEqual(data);
  });

  it("sets isLoading to false", async () => {
    fetch.mockResponseOnce(JSON.stringify([]));
    renderer.render(
      <DataFetcher endpoint="/some-endpoint" render={jest.fn()} />
    );
    const instance = renderer.getMountedInstance();
    await instance.componentDidMount();

    expect(instance.state.isLoading).toBe(false);
  });
});
