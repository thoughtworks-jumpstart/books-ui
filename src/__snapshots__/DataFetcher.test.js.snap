// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when data is not loaded shows a loading message 1`] = `
<div>
  Loadingâ€¦
</div>
`;
